lib/collection/src/shards/replica_set/update.rs

```diff
@@ -130,80 +131,80 @@ impl ShardReplicaSet {
-            let local = self.local.read().await;
-            let this_peer_id = self.this_peer_id();
-
-            // target all remote peers that can receive updates
-            let active_remote_shards: Vec<_> = remotes
-                .iter()
-                .filter(|rs| self.peer_is_active_or_pending(&rs.peer_id))
-                .collect();
-
-            // local is defined AND the peer itself can receive updates
-            let local_is_updatable =
-                local.is_some() && self.peer_is_active_or_pending(&this_peer_id);
-
-            if active_remote_shards.is_empty() && !local_is_updatable {
-                return Err(CollectionError::service_error(format!(
-                    "The replica set for shard {} on peer {} has no active replica",
-                    self.shard_id, this_peer_id
-                )));
-            }
+        let remotes = self.remotes.read().await;
+        let local = self.local.read().await;
+        let mut clock = self.clock_set.lock().await.get_clock();
 
-            let operation = OperationWithClockTag::from(operation); // TODO: Assign clock tag!
+        let this_peer_id = self.this_peer_id();
 
-            let mut update_futures = Vec::with_capacity(active_remote_shards.len() + 1);
+        // target all remote peers that can receive updates
+        let active_remote_shards: Vec<_> = remotes
+            .iter()
+            .filter(|rs| self.peer_is_active_or_pending(&rs.peer_id))
+            .collect();
 
-            if let Some(local) = local.deref() {
-                if self.peer_is_active_or_pending(&this_peer_id) {
-                    let local_wait =
-                        if self.peer_state(&this_peer_id) == Some(ReplicaState::Listener) {
-                            false
-                        } else {
-                            wait
-                        };
+        // local is defined AND the peer itself can receive updates
+        let local_is_updatable = local.is_some() && self.peer_is_active_or_pending(&this_peer_id);
 
-                    let operation = operation.clone();
+        if active_remote_shards.is_empty() && !local_is_updatable {
+            return Err(CollectionError::service_error(format!(
+                "The replica set for shard {} on peer {} has no active replica",
+                self.shard_id, this_peer_id
+            )));
+        }
 
-                    let local_update = async move {
-                        local
-                            .get()
-                            .update(operation, local_wait)
-                            .await
-                            .map(|ok| (this_peer_id, ok))
-                            .map_err(|err| (this_peer_id, err))
-                    };
+        let current_clock_tick = clock.tick_once();
+        let clock_tag = ClockTag::new(this_peer_id, clock.id() as _, current_clock_tick);
+        let operation = OperationWithClockTag::new(operation, Some(clock_tag));
```

Comment:
What about an `Operation::with_clock_tag(clock_tag)` function that constructs the `OperationWithClockTag` for us?

Just a style change, but I guess it may be a nice one.

Reply:
I did implement this now. I also removed

Reply:
I did implement this now. I also removed `From<Operation>` implementation which silently wrapped the type without clock tag, which may go unnoticed and be undesirable in some cases.

This changes makes it more explicit, and shorter. :smile:

Reply:
I did implement this now. I also removed `From<Operation>` implementation which silently wrapped the type without clock tag, which may go unnoticed and be undesirable in some cases.

This changes makes it more explicit, and shorter. :smile:

Reply:
I reverted this as well. Seems a bit too much to add and export a trait just to construct `OperationWithClockTag`.

I've also tried adding explicit methods `OperationWithClockTag::with_tag`/`OperationWithClockTag::without_tag`, but, IMO, these are also annoying to use. :/

For now, I simply reverted to using `From`, but if you're strongly against it we can add `with_tag`/`without_tag`.